libvirt/tests/src/cpu/diagnose_9c_forward.py

```
@@ -0,0 +1,129 @@
+    """
+    Starts spin lock torture test to cause spinlocks in guest
+
+    :param session: guest session
+    """
+    logging.debug("session is %s" % session)
+    cmd = "modprobe locktorture torture_type=spin_lock"
+    err, out = cmd_status_output(cmd, shell=True, session=session)
+    if err:
+        raise OSError("Could not start locktorture. %s" % out)
+
+
+def confirm_forwarding(hz):
+    """
+    Confirm if forwarding is happening or not depending on set
+    hz parameter
+
+    :param hz: if 0 expect no forwarding, if positive expect forward
+    """
+
+    if int(hz) > 0:
+        wait_for_forward_count_to_be_non_zero(FORWARD_COUNTER_TIMEOUT)
+    else:
+        confirm_forward_count_is_still_zero_after(CHECK_INTERVAL)
+
+
+def confirm_forward_count_is_still_zero_after(interval):
+    """
+    Wait for a certain time interval and then confirm
+    that the counter didn't go up
+    """
+    time.sleep(interval)
+    err, out = get_forward_value()
+    if err or int(out) != 0:
+        raise TestFail("Counter wasn't zero although forward"
+                       " was inactive. Value is %s" % out)
+
+
+def get_forward_value():
+    """
+    Reads sysfs for forward value
+    """
+    cmd = "cat /sys/kernel/debug/kvm/diag_9c_forward"
+    return cmd_status_output(cmd, shell=True)
+
+
+def wait_for_forward_count_to_be_non_zero(timeout):
+    """
+    Wait for the forward counter to become > 0. Fail if it doesn't
+    after timeout.
+
+    :param timeout: maximum time to wait for counter to go up
+    """
+
+    def count_is_non_zero():
+        """ Confirm if the count is non-zero """
+        err, out = get_forward_value()
+        if err:
+            return False
+        elif int(out) <= 0:
+            return False
+        elif int(out) > 0:
+            return True
+
+    is_non_zero = wait_for(count_is_non_zero, step=10, timeout=timeout)
+    if not is_non_zero:
+        raise TestFail("Counter didn't go up though forwarding active.")
```

As a common rule, we do use test.fail() instead of raising TestFail exception directly in this repo.

---

@dzhengfy I prefer to raise the exception here directly because then I don't have to pass `test` to the top level function and the top level function can be more easily moved, e.g. to avocado-vt if found useful for other tests. What do you think?